### PR TITLE
Zero supression detection 

### PR DESCRIPTION
**Pull request by kasbah**
Hey, was motivated to work on this after it arose again in https://github.com/mcous/gerber-to-svg/issues/37. I noticed you opened this as #15 on here. 

Just started by writing a test. Does the test make sense? Any guidance appreciated.

### CODE REVIEW

File: test/gerber-parser_drill_test.js
Comment (mcous) [inline test/gerber-parser_drill_test.js]: gerber-parser knows how to detect this, so we may wanna leave it out of this test

File: test/gerber-parser_drill_test.js
Comment (mcous) [inline test/gerber-parser_drill_test.js]: I think it also knows how to detect decimals? So we'd want something like `X7550Y14000\n`

File: test/gerber-parser_drill_test.js
Comment (kasbah) [reply test/gerber-parser_drill_test.js]: I was just trying to _only_ have the warnings I care about otherwise I have to make how many other warnings there are part of the test, or do something more complicated.

Comment (mcous) [reply test/gerber-parser_drill_test.js]: Oh I see! Carry on then. Could use expect.satisfy... but it gets pretty verbose. Maybe it's time to switch this codebase over to ES6ish to get some arrow functions?

``` js
expect(warnings).to.satisfy(function (warns) {
  return warns.some(function (w) {
    return /detected leading/.test(w.message)
  })
})
```

File: test/gerber-parser_drill_test.js
Comment (kasbah) [reply test/gerber-parser_drill_test.js]: Sorry, I don't really know what I am doing. So what I wrote actually has a leading zero but the decimal dots are not needed (but don't hurt)?

File: test/gerber-parser_drill_test.js
Comment (kasbah) [reply test/gerber-parser_drill_test.js]: I actually didn't know about `to.satisfy` so that's probably better.

Comment (kasbah) [reply test/gerber-parser_drill_test.js]: The order is kind of important to the warnings as well i.e. assume "trailing" then detect "leading".

File: test/gerber-parser_drill_test.js
Comment (mcous) [reply test/gerber-parser_drill_test.js]: When converting coordinate strings to numbers, if there's a decimal point in the string it will just use that and ignore any suppression setting: https://github.com/mcous/gerber-parser/blob/master/lib/normalize-coord.js#L27

So I guess if all the coordinates were decimal coordinates, we would render properly, but the message about detecting "leading" or "trailing" would be inaccurate because that's not what's there

File: test/gerber-parser_drill_test.js
Comment (mcous) [reply test/gerber-parser_drill_test.js]: I agree that the order is kind of important, but is it important enough to test for in this case? I would probably skip it given it's pretty much covered elsewhere, but I've got no strong feelings either way

**Comment by kasbah** [comment]
So, in order to switch the way coordinates are parsed on the fly I propose a new type of output object

```js
{
  type: 'reset',
  line: GERBER_LINE_NO
}
```

And then we either transform the existing coordinates (my preference) or implement some sort of [rewindable stream](https://stackoverflow.com/questions/36980201/how-to-reset-nodejs-stream/40865419#40865419). What do you think?

Edit: For the former, a coordinate object could have an additional `source` field so we can re-parse it if needed. 

Edit2: We have to be careful of not allowing the parser to be put into an infinite loop (continually resetting) with conflicting coordinates.

**Comment by mcous** [comment]
So that was my initial thought, too, but I don't think I like the idea of the stream consumer needing to care about internal parsing stuff.

For file type detection, we concatenate the chunks to a string at `this._stash` until we know the type, and then replay parsing from the stash. Perhaps we could have a similar stash of coordinate objects with all relevant information that we hold off on transforming and emitting until we detect a suppression format or hit some arbitrary line limit?

Aside: we should also look into detecting coordinate format. e.g. if we assume 2:4 and then see a coordinate string that is 7 characters long, we probably know we're doing 2:5 instead.

**Comment by kasbah** [comment]
So just to go a bit overkill on illustration here:



The only (non?) issue is we have to read all the coordinates if no suppression is specified and none is detected.

**Comment by mcous** [comment]
Yup, that makes perfect sense. I think your idea of storing an object with the string itself makes the most sense for the stash. In terms of replay we're probably only talking like, I dunno, 100 elements at most? That's probably fine. There are a lot of things we could do elsewhere to speed things up, anyway. That's a separate discussion, though

File: test/parse-coord_test.js
Comment (kasbah) [inline test/parse-coord_test.js]: @mcous will refactor the module layout after but do these tests make sense?

File: test/parse-coord_test.js
Comment (kasbah) [inline test/parse-coord_test.js]: @mcous will refactor the module layout after but do these tests make sense?

Comment (mcous) [reply test/parse-coord_test.js]: Yup this seems straightforward

**Comment by kasbah** [comment]
Ok, this works but needs refactoring still and i'll get the coverage back up but some questions for you:

1. Do you mind the way that it's `parseCoord.detectZero` or should I refactor the module structure or make another module? Planning to add `parseCoord.detectPlaces` like this as well if not. 
2. You said `this._stash` might be a good pattern, what did you mean by `this` though? I have put it in a module level `stash` variable for now, any objections?
3. As it is it requires `M30` or `M00` to recognize that there are no more coordinates to try and detect from, are there any issues with this?

Edit: I think I need to change this to stash tool changes as well, right?

File: lib/_parse-drill.js
Comment (mcous) [inline lib/_parse-drill.js]: `lodash.partial` is a `devDependency`, not a `dependency`. I generally avoid lodash modules in actual code (as much as possible, at least) because they can be pretty big.

File: lib/_parse-drill.js
Comment (mcous) [inline lib/_parse-drill.js]: It's ok just to use a normal anonymous function instead of lodash.partial here

File: lib/_parse-drill.js
Comment (mcous) [inline lib/_parse-drill.js]: This definitely won't work because this stash array will be shared by _all_ parsers that are created. When I referenced `this._stash` this refers to `parser`. See the parser module itself: https://github.com/mcous/gerber-parser/blob/master/lib/parser.js#L44

File: lib/parse-coord.js
Comment (mcous) [inline lib/parse-coord.js]: I think these can be combined into just two regexs:

``` js
var RE_TRAILING = /[XY]0\d+/
var RE_LEADING = /[XY]\d+0(?=\D|$)/
```

Also can you move these constants up to the top of the file?

**Review by mcous — CHANGES_REQUESTED** [review]
Gonna hold off on going too deep into the review, because I think we might still be a little in flux here.

> Do you mind the way that it's parseCoord.detectZero or should I refactor the module structure or make another module?

I'm not in love with it. Could we just turn the export into namespace object with the functions (including the original export) as methods?

> You said this._stash might be a good pattern, what did you mean by this though? I have put it in a module level stash variable for now, any objections?

Yup, I object! Module level stash will be shared by all parsers, which is bad™. `this` refers to `parser` in the context of `parse-drill` and `parse-gerber`.

> As it is it requires M30 or M00 to recognize that there are no more coordinates to try and detect from, are there any issues with this?

Not necessarily, but I like to keep the drill parser as lenient as possible because you never know what a CAD program is going to output. It would be better to just use the actual end of the file somehow.

> I think I need to change this to stash tool changes as well, right?

Yeah, while we're trying to infer format we pretty much have to be stashing everything.

**Comment by kasbah** [comment]
>>As it is it requires M30 or M00 to recognize that there are no more coordinates to try and detect from, are there any issues with this?

>Not necessarily, but I like to keep the drill parser as lenient as possible because you never know what a CAD program is going to output. It would be better to just use the actual end of the file somehow.

We'd have to implement [`transform._flush`](https://nodejs.org/api/stream.html#stream_transform_flush_callback) and add a `flush` method to `_parse-drill` that handles it.

**Comment by kasbah** [comment]
So this now passes tests again and I am generally happy with it though I would make two changes. 

1. I would make the gerber parser mirror the drill parser more by passing the line number through the parse function as well, getting rid of the `-1` initialisation of the commands and [this special case](https://github.com/mcous/gerber-parser/pull/17/files#diff-3fab227e34d65fe9bb2d6ea53eec41cfR90)
2. rename all `parseXXX.parse` to `xxxParser.parse` (and the filenames appropriately)

Wanted to get feedback on the general direction first though. Let me know what you think @mcous.

**Comment by kasbah** [comment]
Just testing this with gerber2svg and it doesn't seem to output any SVGs anymore.  :open_mouth:

**Comment by kasbah** [comment]
Ok, the lack of calling the callback in `_flush` was the problem. Seems to work now.

**Comment by mcous** [comment]
Ah sorry for the delay on this! Got real busy out of nowhere.

> I would make the gerber parser mirror the drill parser more by passing the line number through the parse function as well, getting rid of the -1 initialisation of the commands and this special case

Absolutely. This has been an annoying discrepancy

> rename all parseXXX.parse to xxxParser.parse (and the filenames appropriately)

The `parseGerber` and `parseDrill` functions are, in fact, functions that are called by the parser itself each time the parser encounters a block. The functions happen to get passed a reference to the parser, but I still think that at the moment they're accurately named. That being said, I'm not married to this architecture so suggestions and ideas for something better are more than welcome

Otherwise, I'll review all this tonight. At the moment with this PR, zero detection is only in place for drill files, correct?

**Comment by kasbah** [comment]
Ok, I will clean up the -1 thing when I get a chance.

>The parseGerber and parseDrill functions are, in fact, functions that are called by the parser itself each time the parser encounters a block. The functions happen to get passed a reference to the parser, but I still think that at the moment they're accurately named. That being said, I'm not married to this architecture so suggestions and ideas for something better are more than welcome

Well we have `parseDrill.parse` and `parseDrill.flush` with my changes so I thought `drillParser.parse` and `drillParser.flush` would be nicer.  

It's only for drill files yes, I didn't know it was necessary for Gerbers. I think doing it for Gerbers and the other format detection is going to be separate PRs.

File: lib/_commands.js
Comment (mcous) [inline lib/_commands.js:4]: I'd rather leave the line numbers out of here. Could we just set `this.line` to 0 or 1 depending on if filetype comes out as gerber or drill (I forget which is which)?

https://github.com/mcous/gerber-parser/blob/master/lib/parser.js#L69

File: lib/_parse-drill.js
Comment (mcous) [inline lib/_parse-drill.js]: As discussed above, let's just keep this as exporting the single parse function for now

File: lib/_parse-drill.js
Comment (mcous) [inline lib/_parse-drill.js]: Could this flush logic be a little more generic and also not in this file (it was already a bit too long)?

I think rather than locking our terminology into just drill files, we can approach this as more of a waiting-for-additional-format-info-stash and have this just live in the parser's flush method.

Obviously we should keep the scope of this to just zero detection in drills for the moment, but if it starts in the parser it'll be easier to tackle the gerber stuff, too.

File: lib/_parse-drill.js
Comment (mcous) [inline lib/_parse-drill.js]: I'm pretty sure this stash logic is too greedy. We'll completely miss stuff `INCH,TZ` where we don't need to do detection anymore. Also, until we see something with a coordinate that we can't parse because of missing format, there's no need to stash anything at all. This explains why the tests need so many `p.end()`s.

**Review by mcous — CHANGES_REQUESTED** [review]
I think this is a good start, but I have a few thoughts looking over this.

I was going to put this in a fancy electronic diagram but then it ended up a little more extensive than I expected...

Here's how I think we should structure the parser logic. I think it should be applicable to both gerber and drill (and more than just zero suppression), but we can definitely keep this as just drill and zeros for the moment:



This logic still has a few holes to fill in (e.g. how do line numbers work in this?), but what do you think? Also it's nearly midnight and my brain is certainly not operating at full capacity, so apologies for that...

**Comment by mcous** [comment]
Shoot, was just re-reading over our comments:

> I would make the gerber parser mirror the drill parser more by passing the line number through the parse function as well, getting rid of the -1 initialisation of the commands and this special case

>> Absolutely. This has been an annoying discrepancy

>>> Ok, I will clean up the -1 thing when I get a chance.

I don't know what I thought I was reading, but I was thinking about the fact that gerber and drill files have line numbers that start at 0 or 1, depending on what type they are. My bad for poor reading comprehension. I was not thinking about the -1 initialization or the fact that the parse functions have no knowledge of line numbers.

File: lib/_parse-drill.js
Comment (kasbah) [reply lib/_parse-drill.js]: Yes, you are right, the `INCH,XX` /`METRIC,XX` parser needs to be part of the detection.

How likely is it that we get a file without any coordinates though? I don't see calling `p.end()` in tests as an issue as it will always be called internally in normal use.

Comment (mcous) [reply lib/_parse-drill.js]: There's no inherent problem with p.end in the tests! It's just a symptom of the overly greedy stash. Just want to make sure we minimize any penalty to the parsing of well-formatted files as we implement this feature to handle poorly formatted ones

**Comment by kasbah** [comment]
I haven't fully wrapped my head around what you have drawn in your graph but my understanding is you'd like the stashing and flushing logic to be more generic. Could you maybe show how you would organize the modules and calling logic differently to what I proposed?

While it makes sense to prepare to add stashing and detection for Gerber files as well I am not sure there is much benefit in complicating the stashing logic by making it lazy (i.e. only stash once the format is needed). Ultimately every file will need a format at some point so we might as well stash and determine it from the start. Well specified files, which put the format up on the top, shouldn't really be affected because we can determine the format right away. 

Regarding passing around the line number, I think initizialising to `-1` and then adding the info on right at the end is quite ugly and would much prefer to pass the line number to the places where the commands are decided upon. 

I think it's hard to get away from passing the line number through like this when we are stashing blocks for detection as we have to be able to "fast forward" the parser with all the correct line numbers. This is why I did it that way.

**Comment by kasbah** [comment]
Hi @mcous, sorry I kind of ran out of steam on this and left it for quite a while. I have just pushed a commit to get zero supression from unit blocks like `INCH,TZ`. You requested two other changes: 

1. **Move the `flush` function out of `_parse-drill.js` and make it more generic**. The way it is at the moment the flush logic is tightly coupled with the detection and it doesn't make much sense to me to move it out of the file. To me it makes sense: the parser can `parse` and the parser can be `flush`ed. The `_drillStash` belongs to the parser and flushing is just a fast forward parsing. 
2. **Don't pass the line number around into the parsing function**. As mentioned previously, we need it to be able to stash and then fast-forward. I am not sure how else to do it. 

I probably don't have time for a larger re-architecture, which I think you are proposing with your drawing? We could maybe leave that for another day and with this we at least have a working implementation. I will test this out on some real world cases in the meantime.

**Comment by mcous** [comment]
> I probably don't have time for a larger re-architecture, which I think you are proposing with your drawing? We could maybe leave that for another day and with this we at least have a working implementation. I will test this out on some real world cases in the meantime.

Yeah, that larger re-architecting would take far more time and energy than I think is really worth it at the moment. I appreciate all the work you've put into this, so let me know how the real world testing goes. Once you're happy we can get this merged in.

**Comment by kasbah** [comment]
I tested it by running gerber2svg, adding some pcb-stackup cases and also by re-building a preview version of kitnic.it with it. A motivating example: [a project with Eagle Gerbers and dodgy drill file that needs detection](http://add-ice40-eink.preview.kitnic.it/boards/github.com/kitnic-forks/ice40_eink_controller/). I haven't spotted any issues so far. 

Should I do anything on these two points or just leave it as is?

> I would make two changes. 
>1. I would make the gerber parser mirror the drill parser more by passing the line number through the parse function as well, getting rid of the `-1` initialisation of the commands and [this special case](https://github.com/mcous/gerber-parser/pull/17/files#diff-3fab227e34d65fe9bb2d6ea53eec41cfR90)
>2. rename all `parseXXX.parse` to `xxxParser.parse` (and the filenames appropriately)

**Comment by kasbah** [comment]
I have switched the [main version of kitnic.it](https://kitnic.it/) to use this branch for now. Ifusb and ice40_fpga_board are both projects that take advantage of the zero detection. So far I have not spotted any issues.

**Comment by kasbah** [comment]
@mcous, can this be merged?

**Comment by mcous** [comment]
Please do! Feel free to use your best judgement moving forward; I'll always do my best to get a review in but you're much closer to this than I am at the moment

**Review by mcous — APPROVED** [review]

**Comment by kasbah** [comment]
Ok, I don't have write access to this repo though. :laughing:

**Comment by mcous** [comment]
Well this is embarrassing...

**Pull request merged**
